rmf_gazebo_plugins/src/readonly.cpp

```diff
@@ -0,0 +1,466 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <gazebo/physics/World.hh>
+#include <gazebo/physics/Model.hh>
+#include <gazebo/physics/Joint.hh>
+#include <gazebo/physics/Link.hh>
+#include <gazebo_ros/node.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/logger.hpp>
+
+#include <memory>
+#include <unordered_map>
+#include <mutex>
+
+#include <gazebo/common/Plugin.hh>
+
+#include <rmf_fleet_msgs/msg/robot_state.hpp>
+#include <building_map_msgs/msg/building_map.hpp>
+#include <building_map_msgs/msg/level.hpp>
+#include <building_map_msgs/msg/graph.hpp>
+
+#include "utils.hpp"
+
+using namespace rmf_gazebo_plugins;
+
+class ReadonlyPlugin : public gazebo::ModelPlugin
+{
+public:
+  using BuildingMap = building_map_msgs::msg::BuildingMap;
+  using Level = building_map_msgs::msg::Level;
+  using Graph = building_map_msgs::msg::Graph;
+  using Location = rmf_fleet_msgs::msg::Location;
+  using Path = std::vector<Location>;
+
+  ReadonlyPlugin();
+  ~ReadonlyPlugin();
+
+  void map_cb(const BuildingMap::SharedPtr msg);
+  void Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf) override;
+  void OnUpdate();
+
+private:
+
+  rclcpp::Logger logger();
+
+  void set_traits();
+  void initialize_graph();
+  void initialize_start(const ignition::math::Pose3d& pose);
+  double compute_ds(const ignition::math::Pose3d& pose, const std::size_t& wp);
+  std::size_t get_next_waypoint(const std::size_t& start_wp,
+      const ignition::math::Vector3d& heading,
+      const ignition::math::Pose3d& pose);
+  Path compute_path(const ignition::math::Pose3d& pose);
+
+  gazebo::event::ConnectionPtr _update_connection;
+  gazebo_ros::Node::SharedPtr _ros_node;
+  gazebo::physics::ModelPtr _model;
+
+  rclcpp::Publisher<rmf_fleet_msgs::msg::RobotState>::SharedPtr robot_state_pub;
```

Comment:
just to follow convention with the rest of the code, I think there should be a prefixed underscore

Reply:
Fixed.